File: sortalgorithm/src/main/java/QuickSort.java
import java.util.Random;

/**
 * 实现快排算法，下面的命名采用算法导论中文版的命名。
 */
public class QuickSort {

    public static void exchange(int[] A, int i, int j) {
        int temp = A[i];
        A[i] = A[j];
        A[j] = temp;
    }

    public static int partition(int[] A, int p, int r) {
        int x = A[r];
        int i = p - 1;
        for (int j = p; j <= r - 1; j++) {
            if (A[j] <= x) {
                i += 1;
                exchange(A, i, j);
            }
        }
        exchange(A, i + 1, r);
        return i + 1;
    }

    public static void quickSort(int[] A, int p, int r) {
        if (p < r) {
            int q = partition(A, p, r);
            quickSort(A, p, q - 1);
            quickSort(A, q + 1, r);
        }
    }

    public static void randomizedQuickSort(int[] A, int p, int r) {
        if (p < r)
        {
            int q = randomizedPartition(A, p, r);
            randomizedQuickSort(A, p, q - 1);
            randomizedQuickSort(A, q + 1, r);
        }
    }

    public static int randomizedPartition(int[] A, int p, int r) {
        // todo: 验证下面一句是否会从 [p, r] 中随机算出一个数
        int i = new Random().nextInt(r - p + 1) + p;
        exchange(A, i, r);
        return partition(A, p, r);
    }
}

File: sortalgorithm/python/quick_sort.py
# coding: utf-8

"""
python is more concise than Java. it's quicker to use python to implement algorithms than use Java.

python implements quick sort algorithm.
"""

import numpy


def swap(arr, i, j):
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp


def quick_sort(arr, left, right):
    # 只有一个元素，立马返回
    if left == right:
        return

    # 有多个元素时，递归求解
    middle_index = int((left + right) / 2)
    middle_element = arr[middle_index]

    i = 0
    j = 0

    while i + j != right + 1 - left:
        if arr[i + left] > middle_element:
            swap(arr, i + left, right - j)
            j += 1
        if arr[i + left] < middle_element:
            i += 1
        if arr[i + left] == middle_element:
            first_middle_element = i + left
            i += 1

    # 注意 middle_element 必须出现在分割点上
    swap(arr, first_middle_element, i + left - 1)

    quick_sort(arr, left, max(i - 2 + left, left))
    quick_sort(arr, min(i + left, right), right)


def test_quick_sort():
    test_arr = numpy.random.rand(10)
    length = len(test_arr)

    quick_sort(test_arr, 0, length - 1)
    for i in range(length - 1):
        assert (test_arr[i] <= test_arr[i + 1])
